src/TableFunctions/TableFunctionS3.cpp

```diff
@@ -35,50 +40,60 @@ void TableFunctionS3::parseArgumentsImpl(const String & error_message, ASTs & ar
+            {"compression_method", {false, [](const String &){ return true; }}}
         };
 
-        std::map<String, size_t> args_to_idx;
-        /// For 4 arguments we support 2 possible variants:
-        /// s3(source, format, structure, compression_method) and s3(source, access_key_id, access_key_id, format)
-        /// We can distinguish them by looking at the 2-nd argument: check if it's a format name or not.
-        if (args.size() == 4)
-        {
-            auto second_arg = args[1]->as<ASTLiteral &>().value.safeGet<String>();
-            if (FormatFactory::instance().getAllFormats().contains(second_arg))
-                args_to_idx = {{"format", 1}, {"structure", 2}, {"compression_method", 3}};
+        std::list<std::pair<String, size_t>> matches;
+        std::set<std::pair<String, size_t>> wrong_matches;
 
-            else
-                args_to_idx = {{"access_key_id", 1}, {"secret_access_key", 2}, {"format", 3}};
-        }
-        /// For 3 arguments we support 2 possible variants:
-        /// s3(source, format, structure) and s3(source, access_key_id, access_key_id)
-        /// We can distinguish them by looking at the 2-nd argument: check if it's a format name or not.
-        else if (args.size() == 3)
+        auto parameters_it = parameters_info.begin();
+        for (size_t index = 1; index < args.size(); )
         {
-            auto second_arg = args[1]->as<ASTLiteral &>().value.safeGet<String>();
-            if (FormatFactory::instance().getAllFormats().contains(second_arg))
-                args_to_idx = {{"format", 1}, {"structure", 2}};
+            const auto & arg = args[index]->as<ASTLiteral &>().value.safeGet<String>();
+            if (wrong_matches.contains({parameters_it->first, index}) && parameters_it->second.second(arg))
+            {
+                /// Valid.
+                matches.push_back({parameters_it->first, index});
+                ++index;
+                ++parameters_it;
+            }
+            else if (parameters_it->second.first && std::next(parameters_it) != parameters_info.end())
+            {
+                /// Invalid and can be skipped.
+                ++parameters_it;
+            }
+            else if (!matches.empty())
+            {
+                /// Invalid and can't be skipped.
+                auto [last_parameter, last_index] = matches.back();
+                parameters_it = parameters_info.find(last_parameter);
+                index = last_index;
+                wrong_matches.emplace(std::move(last_parameter), last_index);
+                matches.pop_back();
+            }
             else
-                args_to_idx = {{"access_key_id", 1}, {"secret_access_key", 2}};
-        }
-        else
-        {
-            args_to_idx = size_to_args[args.size()];
+            {
+                /// No possible combinations left.
+                throw Exception(ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH, error_message);
+            }
```

Comment:
I think, we should done it simpler, because now it's hard to understand what's going on here. Also, looks like it works wrong according to tests. I suggest just to find position of `format` parameter and then it'll be easy to parse parameters related to auth to the left from it and parameters related to format to the right.

Reply:
I doubt it will be simpler, at least for reading. I did try to make it as simple as possible, having most general approach to the problem, even sacrificing performance (this is O(N^2) code) and I did localize all special conditions in `parameters_info` map, and the other part is actually pretty compact and general. The reason it has fails actually is my typo when I fixed `count` -> `contains` thing. What if I just fix the bugs and add detailed comments on that? We could make it ideally simple in future as soon as we will have keyword arguments for functions.

Reply:
Ok, please, write a comment.